src/storage/mvcc/reader/forward_scanner.rs

```diff
@@ -235,20 +235,53 @@ impl<S: Snapshot> ForwardScanner<S> {
                 (Key::from_encoded_slice(res.0), res.1, res.2)
             };
 
-            // Attempt to read specified version of the key. Note that we may get `None`
-            // indicating that no desired version is found, or a DELETE version is found, or
-            // when `has_write == false`. We may also get `Err` due to key lock. We need to
-            // ensure that the scanner keeps working for future calls even when meeting key
-            // lock errors (why??), so we don't apply `?` operator to the result here thus
-            // iterators can be moved forward.
-            let result = self.get(&current_user_key, has_write, has_lock);
+            // `result` stores intermediate values, including errors. When there is an error, for
+            // example, key locked error, we need to do the rest of the work like stepping cursors
+            // instead of stop there.
+            let mut result = Ok(None);
+
+            // `get_ts` is the real used timestamp. If user specifies `MaxInt64` as the timestamp,
+            // we need to change it to a most recently available one.
+            let mut get_ts = self.ts;
+
+            // `met_next_user_key` stores whether the write cursor has been already pointing to
+            // the next user key. If so, we don't need to compare it again when trying to step
+            // to the next user key later.
+            let mut met_next_user_key = false;
 
-            if has_write {
-                self.move_write_cursor_to_next_user_key(&current_user_key)?;
-            }
             if has_lock {
+                match self.isolation_level {
+                    IsolationLevel::SI => {
+                        assert!(self.lock_cursor.valid());
+                        let lock_result = {
+                            let lock_value = self.lock_cursor.value(&mut self.statistics.lock);
+                            Lock::parse(lock_value)
+                        };
+                        match lock_result.and_then(|lock| {
+                            super::util::check_lock(&current_user_key, self.ts, &lock)
```

Comment:
Why not return error directly when check_lock failed.

Reply:
we need to step the write_cursor next in order to scan more later